lib/MusicBrainz/Server/Data/ArtistCredit.pm

```diff
@@ -25,26 +27,31 @@ sub get_by_ids
                 "JOIN artist ON artist.id=artist_credit_name.artist " .
                 "WHERE artist_credit IN (" . placeholders(@ids) . ") " .
                 "ORDER BY artist_credit, position";
+
     my %result;
     my %counts;
-    foreach my $id (@ids) {
-        my $obj = MusicBrainz::Server::Entity::ArtistCredit->new(id => $id);
-        $result{$id} = $obj;
-        $counts{$id} = 0;
-    }
     for my $row (@{ $self->sql->select_list_of_hashes($query, @ids) }) {
-        my %info = (
-            artist_id => $row->{artist},
-            name => $row->{ac_name}
-        );
-        $info{join_phrase} = $row->{join_phrase} // '';
-        my $obj = MusicBrainz::Server::Entity::ArtistCreditName->new(%info);
-        $obj->artist($self->c->model('Artist')->_new_from_row($row));
         my $id = $row->{artist_credit};
-        $result{$id}->add_name($obj);
+        $counts{$id} //= 0;
+        $result{$id} //= MusicBrainz::Server::Entity::ArtistCredit->new(id => $id);
+
+        my $acn = MusicBrainz::Server::Entity::ArtistCreditName->new(
+                      artist_id => $row->{artist},
+                      artist => $self->c->model('Artist')->_new_from_row($row),
+                      name => $row->{ac_name},
+                      join_phrase => $row->{join_phrase} // '',
+                  );
+        $result{$id}->add_name($acn);
         $counts{$id} += 1;
     }
     foreach my $id (@ids) {
+        if (!defined $counts{$id}) {
+            # It's unclear how this could happen, but it seems to be the
+            # cause of MBS-8806. Perhaps we'll find out with the call trace
+            # provided by cluck.
+            cluck "Attempted to load non-existent AC $id, please investigate";
+            next;
```

Comment:
`cluck` will not die, it only outputs a trace, so `next` explicitly skips to the next iteration.

I really have no idea how it is possible that the method is called with a soon-to-be-but-not-yet ID. It somehow must leak from the transaction creating a new AC, before it commits, but I don’t see how.